### PR TITLE
[Non-record] Scaled Byte-level H-Net matches 4-hour subword-level baseline (H-Net val_bpb = 1.2070)

### PR DESCRIPTION
**Pull request by DariusFeher**
## Summary

>**12L H-Net (3 Encoder + 6 Main + 3 Decoder) + byte260 + GQA KV4 + INT6 GPTQ + QAT + zstd-22 + Stride-64 Sliding Eval; val_bpb: 1.2070**

Follow-up to [PR #1104](https://github.com/openai/parameter-golf/pull/1104), which introduced a byte-level H-Net that learns whitespace-aligned, word-like boundaries from raw bytes but reached only 1.36 BPB on a 4 hours run, significantly worse than the competition's 4 hours baseline of 1.20.

This submission closes that gap. By scaling the architecture from 9 layers (2 outer layers x2; Encoder + Decoder) to 12 layers (3 outer layers x2; Encoder + Decoder), adding INT6 GPTQ quantization, quantization-aware training (QAT), sliding window eval, and a `torch.compile` compatible chunking strategy, the H-Net `byte260` model now reaches **1.2070 BPB** in a 4-hour set-up, matching the 4-hour baseline (1.2074) and a comparable `sp1024` H-Net (1.2107).

We also include a detailed comparison of `byte260` vs `sp1024` H-Net, analysis of how the fixed chunking strategy impacts sliding window eval, and results on dynamic chunking (rather than fixed cap chunking) at inference (see README for more details).

#### 4 hours runs

| Config | BPB | Steps | Artifact size |
|--------|-----|------:|--------:|
| **H-Net `byte260` 12L (3 OLs, KV=4)** | **1.2070** | 153,381 | 15.8 MB |
| H-Net `sp1024` 11L (3 OLs, KV=4, div=2) | 1.2107 | 84,278 | 14.6 MB |
|**Baselines**||
| 4-hour baseline | 1.2074 | 329,430 | 15.8 MB |
| H-Net `byte260` 9L ([PR #1104](https://github.com/openai/parameter-golf/pull/1104)) | 1.3595 | 85,242 | 16.0 MB |

#### 10 min runs

| Config | BPB | Steps | Artifact size |
|--------|-----|------:|--------:|
| H-Net `byte260` 12L (3 OLs, KV=4) chunk_size=6 | 1.3123 | 5,957 | 15.2 MB |
| H-Net `sp1024` 11L (3 OLs, KV=4) chunk_size=6 | 1.2754 | 6,161 | 14.1 MB |
|**Baselines**||
| Naive baseline (9L 512dim `sp1024` KV4) | 1.2244 | 13,780 | 15.9 MB |
| H-Net `byte260` 9L OL2 chunk_size=9 ([PR #1104](https://github.com/openai/parameter-golf/pull/1104)) | 1.4116 | 4,520 | 15.8 MB |
| H-Net `sp1024` 9L OL2 chunk_size=12 ([PR #1104](https://github.com/openai/parameter-golf/pull/1104)) | 1.3734 | 4,466 | 16.0 MB |

(*OL = Outer Layers, used for H-Net Encoders and Decoders*)

`sp1024` H-Net represents the subword-level H-Net. For more details about chunk_divisor, see README

## Changes compared to [PR #1104](https://github.com/openai/parameter-golf/pull/1104)

| | #1104 | This PR |
|--|-------|---------|
| Architecture | 9L, OL2 | 12L, OL3 |
| Parameters | 17.5M | 23.0M |
| Quantization | int8+zlib | int6 GPTQ + zstd-22 |
| QAT | None | 85% of warmdown |
| Evaluation | Roundtrip | Sliding window (stride 64) |
| Best 4hr BPB | 1.3595 | **1.2070** |

## Architecture

Same 1-stage H-Net layout as [PR #1104](https://github.com/openai/parameter-golf/pull/1104), scaled up:

```
Input -> Embedding -> Encoder (3 blocks) -> Routing -> ChunkLayer (L -> C)
      -> Main Transformer (6 blocks) -> DeChunkLayer (C -> L)
      -> + Residual Skip -> Decoder (3 blocks) -> LM Head
```

- **12 layers total**: 3 encoder + 6 main + 3 decoder (OL3)
- **512 model dim**, 8 heads, 4 KV heads (GQA)
- **22.97M parameters**, `byte260` tokenizer (vocab=260)
- **Chunk target size**: 6

## Additional findings / analysis (see README for  details)

- At 10 minutes, `byte260` still has a significant gap compared to `sp1024` (+0.037 BPB) With a 4-hour budget, `byte260` closes the gap, matching both the `sp1024` H-Net (1.2070 `byte260` vs 1.2107 `sp1024`) and the naive baseline (1.2074).

- The byte H-Net still has clear optimization headroom as val BPB is still decreasing at the 4 hours cutoff. This suggests that part of the remaining gap is due to optimization budget rather than a hard limit of byte-level H-Net.

- To enable torch.compile, we choose a simple solution of capping the number of boundaries at a fixed value per batch.  If the router predicts more than this cap, excess boundaries at the tail are silently dropped. Since sliding window eval only scores the tail of the sequence, this directly degrades sliding eval results. We analyze this effect and a fix in the README.

- Models trained with a fixed boundary cap transfer well to dynamic chunking at inference (no cap). Some see significant improvements, especially those that were truncated during training (see `sp1024` with chunk_divisor=4), and none show any degradation. Removing the cap at inference restores the dropped boundaries, but can't undo the training-time 'damage'.

- Both `byte260` and `sp1024` use shared hyperparameters (except number of layers). Per-model tuning of QAT and chunking settings would likely improve both.

## Reproduction

```bash
# Best byte260 4-hour run (1.2070 BPB)
COMPILE_MODEL=1 OUTER_LAYERS=3 NUM_LAYERS=12 MODEL_DIM=512 \
    TRAIN_SEQ_LEN=2048 MAX_WALLCLOCK_SECONDS=14400 ITERATIONS=500000 \
    RATIO_LOSS_WEIGHT=0.05 LATE_QAT_THRESHOLD=0.85 WARMDOWN_ITERS=25000 \
    EMA_DECAY=0 EVAL_STRIDE=64 EVAL_BATCH_SEQS=16 \
    GPTQ_RESERVE_SECONDS=120 GPTQ_CALIB_BATCHES=512 PRUNE_PCT=0.03 \
    CHUNK_DIVISOR=4 CHECKPOINT_EVERY=50000 \
    MODEL_SAVE_PATH=models/byte260_compile_12L_OL3_4hr_qat85.pt \
    RUN_ID=byte260_compile_12L_OL3_4hr_qat85 \
    torchrun --standalone --nproc_per_node=8 hnet/train_gpt_hnet_compile_gptq.py

# 10 min byte260 equivalent
COMPILE_MODEL=1 OUTER_LAYERS=3 NUM_LAYERS=12 MODEL_DIM=512 \
    TRAIN_SEQ_LEN=2048 MAX_WALLCLOCK_SECONDS=600 ITERATIONS=20000 \
    RATIO_LOSS_WEIGHT=0.05 LATE_QAT_THRESHOLD=0.15 EMA_DECAY=0 \
    EVAL_STRIDE=64 EVAL_BATCH_SEQS=16 \
    GPTQ_RESERVE_SECONDS=45 GPTQ_CALIB_BATCHES=256 PRUNE_PCT=0.03 \
    CHUNK_DIVISOR=4 \
    MODEL_SAVE_PATH=models/sweep_div4_compile_10min.pt \
    RUN_ID=sweep_div4_compile_10min \
    torchrun --standalone --nproc_per_node=8 hnet/train_gpt_hnet_compile_gptq.py
```

> *Note*: Both experiments use `seed=1337`

## Compliance

- [x] Artifact ≤ 16 MB (15.8 MB)
- [x] 8×H100 training
- [x] No training on validation data
- [x] No network calls during evaluation
- [x] Non-record: extended run (153k steps / 4h)

## Credits

- Hwang et al. (2025), [*Dynamic Chunking for End-to-End Hierarchical Sequence Modeling*](https://arxiv.org/abs/2507.07955). Official code: [goombalab/hnet](https://github.com/goombalab/hnet)
- PRs with QAT / GPTQ / INT6 quantization / sliding window eval

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Non-record] Scaled Byte-level H-Net matches 4-hour subword-level baseline (H-Net val_bpb = 1.2070)

**BPB:** 1.2070 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `1b7a4ac4d626`, file `records/track_non_record_16mb/2026-04-01_Scaled_HNet_Byte260_and_Sp1024/train_gpt_byte_hnet.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=260, code=84058 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=260, code=84058 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
